### PR TITLE
Core/Packets: Port packet handling rework from 6.x branch

### PR DESCRIPTION
**Pull request by jackpoz**
**Changes proposed:**
-  Port packet handling reword from 6.x branch. Old handlers are still used at the moment, these changes allow to port old handlers to new packet-specifc ones.
- Opcode.cpp had some weird setting, like SMSG_ with Handle_Null handler of MSG_ with Handle_ServerSide. There were also few mismatches between the Opcodes enum name and the string description in the table.

**Target branch(es):** 3.3.5/6.x
3.3.5

**Issues addressed:**
No issues

**Tests performed:** (Does it build, tested in-game, etc)
Logged ingame with a character, I could jump, cast spell, invite other players.
Send a SMSG_ACCOUNT_DATA_TIMES from BotFarm and worldserver showed the expected error message "Received not allowed opcode"

**Known issues and TODO list:**

The point of this PR is NOT to handle ANY opcode with the new handlers but just to add the infrastructure to make it possible. All opcodes handlers will have to be reworked later (with small commits/small PRs only please!)

### CODE REVIEW

File: src/server/game/Server/Protocol/Opcodes.h [hunk 1]
```diff
@@ -1342,6 +1349,9 @@ enum Opcodes
     NUM_MSG_TYPES                                   = 0x51F
 };
 
+typedef Opcodes OpcodeClient;
```
Comment (jackpoz) [inline src/server/game/Server/Protocol/Opcodes.h]: This allows to have same code as 6.x while keeping only 1 opcodes enum (at the price of opcode type-safety)

File: src/server/game/Grids/Notifiers/GridNotifiersImpl.h [hunk 2]
```diff
@@ -526,8 +526,6 @@ void Trinity::LocalizedPacketDo<Builder>::operator()(Player* p)
 
         i_builder(*data, loc_idx);
 
-        ASSERT(data->GetOpcode() != MSG_NULL_ACTION);
```
Comment (jackpoz) [inline src/server/game/Grids/Notifiers/GridNotifiersImpl.h]: An error is logged in packet sender if NULL_OPCODE is used

File: src/server/game/Server/Packets/AllPackets.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef AllPackets_h__
+#define AllPackets_h__
```
Comment (jackpoz) [inline src/server/game/Server/Packets/AllPackets.h]: No packets are implemented with the new handlers at the moment

File: src/server/game/Server/Protocol/Opcodes.h [hunk 4]
```diff
@@ -1365,30 +1375,92 @@ class WorldPacket;
...
+            return _internalTableClient[index];
+        }
+
+    private:
+        template<typename Handler, Handler HandlerFunction>
+        void ValidateAndSetClientOpcode(OpcodeClient opcode, char const* name, SessionStatus status, PacketProcessing processing);
+
+        void ValidateAndSetServerOpcode(OpcodeServer opcode, char const* name, SessionStatus status);
+
+        ClientOpcodeHandler* _internalTableClient[NUM_OPCODE_HANDLERS];
```
Comment (jackpoz) [inline src/server/game/Server/Protocol/Opcodes.h]: only ClientOpcodeHandle are stored, SMSG are stored in same table with Handle_ServerSide handler

File: src/server/game/Server/WorldSocket.cpp [hunk 5]
```diff
@@ -1,20 +1,20 @@
...
-*
-* This program is distributed in the hope that it will be useful, but WITHOUT
-* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
-* FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
-* more details.
-*
-* You should have received a copy of the GNU General Public License along
-* with this program. If not, see <http://www.gnu.org/licenses/>.
-*/
+ * Copyright (C) 2008-2016 TrinityCore <http://www.trinitycore.org/>
```
Comment (jackpoz) [inline src/server/game/Server/WorldSocket.cpp]: WinMerge showed some annoying diff here with the missing space

File: src/server/game/Server/Protocol/Opcodes.h [hunk 6]
```diff
@@ -24,11 +24,18 @@
...
 #include "Common.h"
+#include <iomanip>
 
-/// List of Opcodes
-enum Opcodes
+enum OpcodeMisc : uint16
+{
+    MAX_OPCODE                                        = 0x3FFF,
+    NUM_OPCODE_HANDLERS                               = (MAX_OPCODE + 1),
+    UNKNOWN_OPCODE                                    = (0x7FFF + 1),
```
Comment (jackpoz) [inline src/server/game/Server/Protocol/Opcodes.h]: to fit UNKNOWN_OPCODE value into uint16 I changed the value from 0xFFFF + 1 to 0x7FFF + 1. I don't see a reason for this to cause any issue.

File: src/server/game/Server/Protocol/Opcodes.h [hunk 7]
```diff
@@ -24,11 +24,18 @@
 #define _OPCODES_H
```
Comment (jackpoz) [inline src/server/game/Server/Protocol/Opcodes.h]: Please not Opcodes.cpp changes are too many to be displayed, check https://github.com/jackpoz/TrinityCore/blob/509fd2d2e4bbfeb22b720224f2c192c44b0666e0/src/server/game/Server/Protocol/Opcodes.cpp

**Comment by Nawuko** [comment]
Awesome 👍

**Comment by Pixel-Pirate** [comment]
that great bro (y)

**Pull request merged**